### DISCUSSION TITLE
Options for splitting a list into two parts?

### DISCUSSION THREAD

**Discussion by JaredDavis22** [tomj74/chunk-templates#43] [Q&A] [answered]
Hi,

I have a list of objects that I need to print on one or two pages.

I would like to have the template filter the given list into two lists. That way the page split logic is all in the template.

given:
class Item {
String getName...
}

List<Item> items
chunk.set("items",items);

Then something like this in the chunk template

pageOneList = $items.stream().filter(i -> c.getName().matches("re for page one goes here"))
pageTwoList =  $items.stream().filter(i -> !c.getName().matches("re for page one goes here"))

I need to test for first and last items in the lists for pageTwoList so running through the same list twice won't work w/o local variable, I think.

I would like something like this: (Assuming I can get the set in the correct spot)

$items|stream(.name, re goes here,true)|set($pageOneList)
$items|stream(.name, re goes here,false)|set($pageTwoList)

true/false = matches or !matches

Any suggestions on paths to investigate?

**Comment by tomj74** [comment, marked as answer]
It looks like you're not doing pagination (eg x items per page), just dividing the list into two lists based on some attribute?

There's (intentionally) a limit to how much logic you can push into the template. Any logic that's sufficiently complex I really would expect to be pre-handled in the java code for your app. For example, pre-split the list and provide the template with the two lists in distinct template vars.

{% loop %} does offer $first and $last and $place tags for your convenience, just set the first_last="true" option

If you really want to do this, you can write your own filter. Take a look at the source code for SliceFilter for an example of a filter that takes a list as input and outputs another list.

Also, please keep in mind that {% exec %} is the intended construct for binding some variables to the value(s) of some expression(s) and then rendering a template, with no need for |set(...) -- you can just make two {% exec %} blocks.

**Comment by JaredDavis22** [comment]
Hi,

Thanks for the response.

Got a code snippet for resolving a tag to text? I don't see where the reflection occurs.

eg given a list entry  and parameter string ".name" how do I do a "chunk approved" method call that eventually calls getName() from the entry?

Object o  = list.get(0) // this is really an Item entry.
String p = ".name"

____

String pValue = somechunkmagic( chunk, o, p);

pValue contains the results of the item.getName() method call.

**Comment by tomj74** [comment]
Ah, good question.  There is a way to hack your list of objects into a list of boxed objects which have a Map interface, which is how chunk simplifies data structures to expose beans/pojos to the templates.

something like this
```
import com.x5.template.Chunk;
import com.x5.template.filters.FilterArgs;
import com.x5.template.filters.ListFilter;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class ListOfObjectsFilter extends ListFilter
{
    public String getFilterName()
    {
        return "sieve";
    }

    @SuppressWarnings("rawtypes")
    public Object transformList(Chunk chunk, List list, FilterArgs arg)
    {
        if (list == null) return list;

        ArrayList names = new ArrayList();
        for (Object listItem : list) {
            Object x = Chunk.boxIfAlienObject(listItem);
            String name = ((Map)x).get("name").toString(); // you will need to handle a classcast exception here!!
            names.add(name);
        }

        return names;
    }
}
```

This just transforms a list of objects into a list of their names, but you can use this strategy to examine any arbitrary field and omit it from the returned list based on matching your arg-provided regex.


UPDATE
here's an alternate way to leverage chunk-ification of your data that might be a bit cleaner (no exceptions to worry about):

```
import com.x5.template.Chunk;
import com.x5.template.ChunkFactory;
import com.x5.template.filters.FilterArgs;
import com.x5.template.filters.ListFilter;
import java.util.ArrayList;
import java.util.List;


public class ListOfObjectsFilter extends ListFilter
{
    public String getFilterName()
    {
        return "sieve";
    }

    @SuppressWarnings("rawtypes")
    public Object transformList(Chunk chunk, List list, FilterArgs arg)
    {
        if (list == null) return list;

        ChunkFactory factory = chunk.getChunkFactory();

        ArrayList names = new ArrayList();
        for (Object listItem : list) {
            Chunk helper = factory.makeChunk();
            helper.set("x", listItem);
            String name = helper.get("x.name");
            names.add(name);
        }

        return names;
    }
}
```
